### DISCUSSION TITLE
Vídeo 07 - Pergunta adicional

### DISCUSSION THREAD

**Discussion by thpoiani** [thpoiani/lab-quarkus#10] [Q&A]
Na nova migration adicionada nesse vídeo, não usei chaves estrangeiras.
Claro que tudo depende muito da natureza dos dados envolvidos, mas você consegue pensar em alguns cenários onde seguir **normalização de bancos de dados**, que realmente é um conceito importante, muito fortemente pode ser uma desvantagem?

**Comment by didifive** [comment]
Na normalização de dados pode ocorrer complexidade de que para conseguir uma informação se precisa lidar com a complexidade de relações das tabelas criadas às vezes tendo que considerar tabelas de relacionamento que só possuem códigos para poder ligar os dados.
Essa complexidade pode trazer dificuldades principalmente no salvamento, atualização e deleção de dados, seja na dificuldade de lidar com a cascata do relacionamento ou mesmo em como tratar os dados órfãos, ou acabar deixando estes dados órfãos que consomem espaço indevidamente.

**Comment by mauromortoni** [comment]
O uso de chaves estrangeiras em bancos de dados relacionais é uma prática comum para garantir a integridade referencial, mas também pode trazer algumas desvantagens. Aqui estão as principais desvantagens de usar chaves estrangeiras:

1. Desempenho
- Overhead em Operações: A verificação de integridade referencial pode adicionar latência a operações de inserção, atualização e exclusão, especialmente em tabelas grandes ou em transações com múltiplas operações.
- Bloqueios: As chaves estrangeiras podem causar bloqueios em tabelas relacionadas, o que pode impactar o desempenho em ambientes de alta concorrência.

2. Complexidade da Manutenção
- Alterações de Schema: A modificação da estrutura do banco de dados pode se tornar mais complexa quando chaves estrangeiras estão em uso, pois podem ser necessárias alterações em várias tabelas relacionadas.
- Cascading Operations: Configurações de exclusão em cascata ou atualização em cascata podem levar a comportamentos inesperados se não forem cuidadosamente gerenciadas.

3. Flexibilidade Limitada
- Rigidez nas Relações: A imposição de chaves estrangeiras pode tornar o banco de dados menos flexível para mudanças futuras, já que as relações entre tabelas são rigidamente definidas.
- Dificuldade em Gerenciar Dados Temporários: Se você estiver lidando com dados temporários ou experimentais, as chaves estrangeiras podem ser um obstáculo.

4. Dependências
- Desacoplamento: A presença de chaves estrangeiras pode criar dependências indesejadas entre tabelas, dificultando o desacoplamento de serviços em arquiteturas de microserviços ou sistemas distribuídos.
- Problemas em Migrações: Durante migrações de dados ou integração com outros sistemas, as chaves estrangeiras podem complicar o processo.

5. Manutenção de Dados
- Dados Orfãos: Se não for configurada corretamente, pode ocorrer a criação de dados órfãos (registros que referenciam um registro que não existe mais), especialmente se as operações em cascata não forem implementadas.
- Consistência de Dados: A manutenção da integridade referencial exige disciplina no design e na implementação do banco de dados, o que pode ser desafiador.

6. Ferramentas e Suporte
- Suporte Limitado em Alguns SGBDs: Algumas soluções de banco de dados NoSQL ou sistemas que não são totalmente relacionais podem não suportar chaves estrangeiras, tornando a integração mais difícil.

7. Aumento do Tempo de Desenvolvimento
- Necessidade de Planejamento: O uso de chaves estrangeiras pode exigir um planejamento mais cuidadoso durante o desenvolvimento do banco de dados, o que pode aumentar o tempo total de desenvolvimento.

Conclusão
Embora as chaves estrangeiras sejam eficazes para manter a integridade referencial, suas desvantagens devem ser consideradas cuidadosamente no design do banco de dados. Em certos casos, pode ser mais apropriado gerenciar a integridade referencial de forma programática ou adotar um design mais flexível, dependendo das necessidades específicas da aplicação e do contexto em que o banco de dados será usado.
